### ISSUE TITLE
[RFC] Crate reorganization 

### ISSUE THREAD

**Issue by vitvakatu** [three-rs/three#82]
# Motivation
Three-rs is a very young project. The major part of it was developed in a very short period of time, that leads us to some structural problems.
The main problem - a lot of things is situated in the root module, while there are more suitable places. 

# New structure
- `camera` module 
    * struct `Camera` _moved from `lib.rs`_
    * trait `Projection`
    * struct `Orthographic`
    * struct `Perspective`
- `input` module - **add documentation**
- `window` module - no changes
- `scene` module
    * `Scene` struct
    * `Background` enum
- `render` struct 
_(?)_ rework `render` method to reduce nesting using dedicated private functions in match statements  
- `macros` module - no changes
- `factory` module
    * struct `Factory`
- `loader` module - **new module** _(?)_
    * struct `Loader`
    `Loader` is stored inside `Factory` struct and provides functions for loading textures, sounds, meshes and so on. This functionality is separated from `Factory`.
- `geometry` module - **new module**
    * struct `Geometry` _moved from `factory.rs`_
    * struct `GeometryShape` _moved from `factory.rs`_
- `mesh` module - **new module**
    * struct `Mesh` _moved from `lib.rs` and `scene.rs`_ 
    * struct `DynamicMesh` _moved from `lib.rs` and `scene.rs`_
    * common trait for both `Mesh` and `DynamicMesh` _(??)_
- `material` module - **new module** 
    * struct `Material` _moved from `scene.rs`_
    * struct `Texture`
- `lights` module - **new module**
    * structs `AmbientLight`, `PointLight`, `HemisphereLight`, `DirectionalLight` _moved from `scene.rs`_
- `node` module - **new module**
    * type `NodePointer`
    * type `Transform`
    * struct `Node`
    * struct `NodeInfo`
    * struct `NodeTransform`
- `object` module - **new module**
    * struct `Object`
    * struct `Group`
- `hub` module - **new module**
    * struct `Hub`
    * all types used by `Hub` and not noticed by me in other modules (`SubNode`, `LightData`, `GpuData`...)
- `sprite` module - **new module**
     * `Sprite` struct
- `controls` module - **new module** - not sure about the name
    * `orbit` submodule - contatins `OrbitalControls`
    * `animation` submodule - contains animation controls

#### Publicity
We should use `pub(crate)` to hide some things form user (not `#[doc(hidden)]`) and to allow access to fields and methods for other modules
#### Documentation
We should remove every `#[allow(missing_docs)]` label from code and never let them in the future.
#### Controls
`Controls` - is a special helper struct with `update` method, that can be used to _control_ different objects on scene. I can see `Controls` being used for different `Camera` modes (orbital camera, FPS camera, free-flight camera, etc) and controlling animations.
So the common way of using it is:
```
let anim = AnimationControls::new(my_object);
anim.settings(...);
...
// somewhere in the main loop:
orbit_camera.update(&win.input);
anim.update(&win.input);
```
I'm not sure the it's good idea though, so give me your opinions, please.

# Summary
10 new modules, major rework in `scene.rs`, `lib.rs` and `factory.rs`. However, I already started to do so in my own branch (not a lot of stuff done though) and I'm sure I can finish the most valuable part at this weekends.

# Possible drawbacks
A lot of relatively small modules, possible problems with interop between different modules (can be solved with `pub(crate)`.

Give me your opinions about this RFC.

**Comment by alteous** [comment]
The proposed crate structure is logical, so +1 from me.

#### Review

> (?) rework render method to reduce nesting using dedicated private functions in match statements

Unnecessary at the moment.

> loader module - new module

Good idea.

> This functionality is separated from Factory.

In general, I'd like to see all the `impl Factory` blocks in one place. Navigation is a nightmare at the moment!

> `GeometryShape`

Could be renamed `geometry::Shape`.

> `struct Texture`

Textures should probably have their own module, since in general they're used for more than just materials.

> hub module - new module

I don't really understand what `Hub` does. It would be helpful to include some notes on its relevance / usage.

> We should use pub(crate) to hide some things form user (not #[doc(hidden)]) and to allow access to fields and methods for other modules

Totally agree.

> Controls

I've no complaints with the current design but it might be worth experimenting with alternatives. Perhaps controls could be toggled on / off for per-frame updates, in the same way meshes can be toggled visible / invisible?

#### Additions

 - Export `camera::Projection`.
 - Remove suffixes like `AmbientLight`, `DirectionalLight` etc. in favour of path prefixed versions, i.e. `light::Ambient`, `light::Directional` etc.

**Comment by vitvakatu** [comment]
>    (?) rework render method to reduce nesting using dedicated private functions in match statements
> Unnecessary at the moment.

Yep, maybe

> Could be renamed geometry::Shape.
> Remove suffixes like AmbientLight, DirectionalLight etc. in favour of path prefixed versions, i.e. light::Ambient, light::Directional etc.
> Export camera::Projection.

Agreed

> Textures should probably have their own module, since in general they're used for more than just materials.

I thought about it, maybe it's better to have dedicated module.

> I don't really understand what Hub does. It would be helpful to include some notes on its relevance / usage.

It's literally the heart of `three-rs`. It controls the scenegraph, and it's the place where `froggy` storage for actual game objects is.

> I've no complaints with the current design but it might be worth experimenting with alternatives. Perhaps controls could be toggled on / off for per-frame updates, in the same way meshes can be toggled visible / invisible?

Perhaps we should discuss it in another issue? The current usage was proposed by @kvark in some way, when I was working on prototype of animation system.

**Comment by alteous** [comment]
> It's literally the heart of three-rs. It controls the scenegraph, and it's the place where froggy storage for actual game objects is.

That makes sense. I seldom see the word 'hub' used. Perhaps it should be renamed `Core` or something similar? Just a thought.

> Perhaps we should discuss it in another issue?

Yes, I think so.

**Comment by alteous** [comment]
#### Extra suggestions

 - Remove `new_` prefix from `Geometry` constructors, replacing `new_box` with `new_cuboid` to avoid keyword clash.
- Rename `Geometry::from_vertices` to `Geometry::with_vertices`.

**Comment by vitvakatu** [comment]
Closed in #86
